aiida_cp2k/workchains/multistage.py

```diff
@@ -0,0 +1,486 @@
+# -*- coding: utf-8 -*-
+"""Multistage workchain"""
+from __future__ import absolute_import
+import os
+from copy import deepcopy
+import six
+import ruamel.yaml as yaml  #does not convert OFF to False
+
+from aiida.common import AttributeDict
+from aiida.engine import append_, while_, WorkChain, ToContext
+from aiida.engine import workfunction as wf
+from aiida.orm import Dict, Int, Float, SinglefileData, Str, RemoteData, StructureData
+from aiida.plugins import CalculationFactory
+
+from aiida_cp2k.workchains import Cp2kBaseWorkChain
+
+Cp2kCalculation = CalculationFactory('cp2k')  # pylint: disable=invalid-name
+
+HARTREE2EV = 27.211399
+
+
+def merge_dict(dct, merge_dct):
+    """ Taken from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
+    Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
+    updating only top-level keys, merge_dict recurses down into dicts nested
+    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
+    ``dct``.
+    :param dct: dict onto which the merge is executed
+    :param merge_dct: dct merged into dct (overwrites dct data if in both)
+    :return: None
+    """
+    import collections
+    for k, _ in merge_dct.items():  # it was .iteritems() in python2
+        if (k in dct and isinstance(dct[k], dict) and isinstance(merge_dct[k], collections.Mapping)):
+            merge_dict(dct[k], merge_dct[k])
+        else:
+            dct[k] = merge_dct[k]
+
+
+@wf
+def merge_Dict(p1, p2):  #pylint: disable=invalid-name
+    """ Overwrite the second Dict into the first Dict """
+    p1_dict = p1.get_dict()
+    p2_dict = p2.get_dict()
+    merge_dict(p1_dict, p2_dict)
+    return Dict(dict=p1_dict).store()
+
+
+def get_kinds_section(structure, protocol_settings):
```

Comment:
Please move it into the `utils` folder. Also consider whether to move into `utils` the functions below.

Reply:
I propose to move all the functions to utils, except for `extract_results` that is very much tailor-made for the multistage workchain and it makes more sense to include it in the same file